File: shannon/minecraft_bot_agent/instantSkills/getBlocksData.js
const fs = require('fs');
const InstantSkill = require('./instantSkill.js');
const { Vec3 } = require('vec3');

class GetBlocksData extends InstantSkill {
    /**
     * @param {import('../types.js').CustomBot} bot
     */
    constructor(bot) {
        super(bot);
        this.skillName = "get-blocks-data";
        this.description = "指定した空間のブロックのデータを取得します";
        this.bot = bot;
        this.params = [
            {
                name: "startPoint",
                type: "vec3",
                description: "始点座標",
                default: "0, 0, 0"
            },
            {
                name: "endPoint",
                type: "vec3",
                description: "終点座標",
                default: "3, 4, 4"
            }
        ]
        this.canUseByCommand = false;
    }

    /**
     * @param {Vec3} startPoint
     * @param {Vec3} endPoint
     */
    async run(startPoint, endPoint) {
        try {
            const blocksData = [];
            const volume = Math.abs((endPoint.x - startPoint.x) * (endPoint.y - startPoint.y) * (endPoint.z - startPoint.z));
            if (volume > 64) {
                return { "success": false, "result": "指定した空間の体積が64を超えています" };
            }

            for (let x = startPoint.x; x <= endPoint.x; x++) {
                for (let y = startPoint.y; y <= endPoint.y; y++) {
                    for (let z = startPoint.z; z <= endPoint.z; z++) {
                        const block = this.bot.blockAt(new Vec3(x, y, z));
                        if (block) {
                            blocksData.push({
                                position: { x, y, z },
                                name: block.name,
                            });
                        }
                    }
                }
            }
            const path = require('path');
            const filePath = path.join(process.cwd(), 'saves', 'minecraft', 'blocks.txt');
            fs.writeFileSync(filePath, JSON.stringify(blocksData, null, 2));
            return { "success": true, "result": `指定した空間のブロックデータを以下に格納しました: ${filePath}` };
        } catch (error) {
            return { "success": false, "result": `${error.message} in ${error.stack}` };
        }
    }
}

module.exports = GetBlocksData;

File: shannon/tools/minecraft/instant_skill/get_blocks_data.py
from langchain.pydantic_v1 import BaseModel, Field
from langchain.tools import BaseTool
from typing import Optional, Type
from langchain.callbacks.manager import (
    AsyncCallbackManagerForToolRun
)
import utils as U


class GetBlocksDataInput(BaseModel):
    # ブロックの情報を取得したい空間の始点座標を指定します。 ex) 0,0,0
    start_point: str = Field(
        description="Specify the start point of the space to get block information. ex) 0,0,0")
    # ブロックの情報を取得したい空間の終点座標を指定します。 ex) 3,4,4
    end_point: str = Field(
        description="Specify the end point of the space to get block information. ex) 3,4,4")


class GetBlocksDataTool(BaseTool):
    # 指定した空間にあるブロックの情報を取得するためのツール
    name = "get-blocks-data"
    description = "Tool to get information about blocks in the specified space in Minecraft. The space volume must be less than or equal to 64. Use the bot's current location coordinates as a reference."
    args_schema: Type[BaseModel] = GetBlocksDataInput

    def _run(
        self, start_point: str, end_point: str, run_manager: Optional[AsyncCallbackManagerForToolRun] = None
    ) -> str:
        """Use the tool synchronously."""
        raise NotImplementedError("This tool requires asynchronous operation.")

    async def _arun(
        self, start_point: str, end_point: str, run_manager: Optional[AsyncCallbackManagerForToolRun] = None
    ) -> str:
        """Use the tool asynchronously."""
        try:
            data = {
                "startPoint": start_point,
                "endPoint": end_point
            }
            import os
            current_dir = os.path.dirname(os.path.abspath(__file__))
            file_path = os.path.join(current_dir, "..", "..", "..", "saves", "minecraft", "blocks.txt")
            response = await U.send_request(
                endpoint='get-blocks-data', data=data, destination="mineflayer", request_type="post")
            if response['success']:
                with open(file_path, 'r') as file:
                    file_content = file.read()
                return f"Success: \n{file_content}"
            else:
                return f"Failed: {response['result']}"
        except Exception as e:
            return f"get-blocks-data: {e}"
